gestalt-asset-core/src/main/java/org/terasology/gestalt/assets/AssetType.java

```diff
@@ -267,33 +274,95 @@ public synchronized void clearProducers() {
     }
 
     /**
-     * Notifies the asset type when an asset is disposed
+     * Notifies the asset type when an asset is created
      *
-     * @param asset The asset that was disposed.
+     * @param asset The asset that was created
      */
-    void onAssetDisposed(Asset<U> asset) {
-        if (asset.getUrn().isInstance()) {
-            instanceAssets.get(asset.getUrn()).remove(new WeakReference<>(assetClass.cast(asset)));
+    synchronized void registerAsset(Asset<U> asset, DisposalHook disposer) {
+        if (closed) {
+            throw new IllegalStateException("Cannot create asset for disposed asset type: " + assetClass);
+        } else {
+            if(addAsset(assetClass.cast(asset))) {
+                references.add(new AssetReference<>(asset, disposalQueue, disposer));
+            }
+        }
+    }
+
+
+    private boolean addAsset(T targetAsset) {
+
+        ResourceUrn targetUrn = targetAsset.getUrn();
+        ResourceUrn parentUrn = targetUrn.getParentUrn();
+
+        Reference<T> reference = loadedAssets.get(parentUrn);
+        T current = reference == null ? null: reference.get();
+        if(current != null){
+            ResourceUrn referenceUrn = current.getUrn();
+            if(!referenceUrn.isInstance() && !targetUrn.isInstance()) {
+                throw new RuntimeException("An already non instance asset has been registered");
+            } else if(targetUrn.isInstance()) { // target object is instanced
+                Asset.AssetNode<Asset<U>> next = current.next;
+                Asset.AssetNode<Asset<U>> newNode = new Asset.AssetNode<>(current, targetAsset);
+                current.next = newNode;
+                newNode.next = next;
+            } else { // target object is non instanced so replace instance with
+                targetAsset.next = new Asset.AssetNode<>(targetAsset, current);
+                Asset.AssetNode<Asset<U>> next = targetAsset.next;
+                while (next != null) {
+                    next.setParent(targetAsset); // update all parents to point to new target
+                    next = next.next;
+                }
+                loadedAssets.put(parentUrn, targetUrn.isInstance() ? new WeakReference<T>(targetAsset) : new SoftReference<T>(targetAsset));
+            }
         } else {
-            loadedAssets.remove(asset.getUrn());
```

Comment:
Removing this statement means loaded assets will contain disposed assets... which means they won't be reloaded automatically if someone requests it again by Urn.